### PR TITLE
State Breaking: Backport fix to prevent failed transactions from commiting state when block gas limit is reached

### PR DESCRIPTION
**Pull request by nddeluca**
Based off of @yihuang's work in https://github.com/cosmos/cosmos-sdk/pull/10814.

### CODE REVIEW

File: baseapp/baseapp.go [hunk 1]
```diff
@@ -616,7 +618,10 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 				panic(sdk.ErrorGasOverflow{Descriptor: "tx gas summation"})
 			}
 		}
-	}()
+	}
+	if mode == runTxModeDeliver {
+		defer consumeBlockGas()
```
Comment (nddeluca) [inline baseapp/baseapp.go:623]: This defer behavior is unchanged and still runs if we do not make it to runMessages for a transaction

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -606,8 +606,10 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 	//
 	// NOTE: This must exist in a separate defer function for the above recovery
 	// to recover from this one.
-	defer func() {
-		if mode == runTxModeDeliver {
+	blockGasConsumed := false
+	consumeBlockGas := func() {
+		if !blockGasConsumed {
```
Comment (nddeluca) [inline baseapp/baseapp.go:611]: This check ensures block gas is only consumed once

File: baseapp/baseapp.go [hunk 3]
```diff
@@ -678,6 +683,8 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 	// Result if any single message fails or does not have a registered Handler.
 	result, err = app.runMsgs(runMsgCtx, msgs, mode)
 	if err == nil && mode == runTxModeDeliver {
+		consumeBlockGas()
```
Comment (nddeluca) [inline baseapp/baseapp.go:686]: This is the main fix -- gas is consumed before calling `msCache.Write()`, ensuring a panic is thrown when if the transaction consumes more than the block limit

**Review by karzak — APPROVED** [review]
Fix looks good, stepped through tests locally as well.

**Pull request merged**